rust/datafusion/src/physical_plan/sql.rs

```diff
@@ -0,0 +1,246 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Execution plan for reading Parquet files
+
+use std::any::Any;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::{fmt, thread};
+
+use super::{RecordBatchStream, SendableRecordBatchStream};
+use crate::error::{DataFusionError, Result};
+use crate::physical_plan::ExecutionPlan;
+use crate::physical_plan::Partitioning;
+use arrow::datatypes::SchemaRef;
+use arrow::error::{ArrowError, Result as ArrowResult};
+use arrow::record_batch::RecordBatch;
+use arrow_sql::postgres::PostgresReadIterator;
+use crossbeam::channel::{bounded, Receiver, RecvError, Sender};
+use fmt::Debug;
+
+use async_trait::async_trait;
+use futures::stream::Stream;
+
+/// Execution plan for scanning a SQL data source
+#[derive(Debug, Clone)]
+pub struct SqlExec {
+    /// SQL connection
+    connection: String,
+    /// SQL query
+    query: String,
+    /// Schema after projection is applied
+    schema: SchemaRef,
+    /// Projection for which columns to load
+    projection: Vec<usize>,
+    /// Batch size
+    batch_size: usize,
+}
+
+impl SqlExec {
+    /// Create a new SQL reader execution plan
+    pub fn try_new(
+        connection: &str,
+        query: &str,
+        projection: Option<Vec<usize>>,
+        batch_size: usize,
+    ) -> Result<Self> {
+        // TODO: we could/should determine the type of database at this point
```

Comment:
I think @andygrove 's suggestion of using some generic database connectivity layer here (e.g. RDBC https://github.com/tokio-rs/rdbc) rather than building our own registry / set of database drivers would be a good idea to avoid duplicating the (substantial) effort needed for various database connectivity.

Reply:
I'll have a look at RDBC's connectivity layer, I wasn't understanding.